File: python/descriptives/stats_descriptives_entropy.py
#!/usr/bin/env python
# coding: utf-8

# # COURSE: Master statistics and machine learning: Intuition, Math, code
# ##### COURSE URL: udemy.com/course/statsml_x/?couponCode=202304 
# ## SECTION: Descriptive statistics
# ### VIDEO: Entropy
# #### TEACHER: Mike X Cohen, sincxpress.com

# In[ ]:


# import libraries
import matplotlib.pyplot as plt
import numpy as np


# In[ ]:


## "discrete" entropy

# generate data
N = 1000
numbers = np.ceil( 8*np.random.rand(N)**2 )
numbers[numbers==7] = 4
plt.plot(numbers,'o')


# In[ ]:


## "discrete" entropy

# generate data
N = 1000
numbers = np.ceil( 8*np.random.rand(N)**2 )


# get counts and probabilities
u = np.unique(numbers)
probs = np.zeros(len(u))

for ui in range(len(u)):
    probs[ui] = sum(numbers==u[ui]) / N

    
# compute entropy
entropee = -sum( probs*np.log2(probs+np.finfo(float).eps) )


# plot
plt.bar(u,probs)
plt.title('Entropy = %g'%entropee)
plt.xlabel('Data value')
plt.ylabel('Probability')
plt.show()


# In[ ]:


## for random variables

# create Brownian noise
N = 1123
brownnoise = np.cumsum( np.sign(np.random.randn(N)) )

fig,ax = plt.subplots(2,1,figsize=(4,6))
ax[0].plot(brownnoise)
ax[0].set_xlabel('Data index')
ax[0].set_ylabel('Data value')
ax[0].set_title('Brownian noise')

ax[1].hist(brownnoise,30)
ax[1].set_xlabel('Data value')
ax[1].set_ylabel('Counts')
plt.show()


# In[ ]:


### now compute entropy
# number of bins
nbins = 50

# bin the data and convert to probability
nPerBin,bins = np.histogram(brownnoise,nbins)
probs = nPerBin / sum(nPerBin)

# compute entropy
entro = -sum( probs*np.log2(probs+np.finfo(float).eps) )

print('Entropy = %g'%entro)


# In[ ]:





# In[ ]:

File: MATLAB/descriptives/stats_descriptives_entropy.m
%%
%   COURSE: Master statistics and machine learning: intuition, math, code										
%      URL: udemy.com/course/statsml_x/?couponCode=202304 
% 
%  SECTION: Descriptive statistics
%    VIDEO: Inter-quartile range
% 
%  TEACHER: Mike X Cohen, sincxpress.com
%

%%

% a clear MATLAB workspace is a clear mental workspace
close all; clear

%% "discrete" entropy

% generate data
N = 1000;
numbers = ceil( 8*rand(N,1).^2 );


% get counts and probabilities
u = unique(numbers);
probs = zeros(length(u),1);

for ui=1:length(u)
    probs(ui) = sum(numbers==u(ui)) / N;
end


% compute entropy
entropee = -sum( probs.*log2(probs+eps) );


% plot
figure(1), clf
bar(u,probs)
title([ 'Entropy = ' num2str(entropee) ])
xlabel('Data value'), ylabel('Probability')

%% for random variables

% create Brownian noise
N = 1123;
brownnoise = cumsum( sign(randn(N,1)) );

figure(1), clf
subplot(211)
plot(brownnoise)
xlabel('Data index'), ylabel('Data value')
title('Brownian noise')

subplot(212)
histogram(brownnoise,30,'Normalization','probability')
xlabel('Data value'), ylabel('Probability')


%% now compute entropy

% number of bins
nbins = 50;

% bin the data and convert to probability
[nPerBin,bins] = histcounts(brownnoise,nbins);
probs = nPerBin ./ sum(nPerBin);

% compute entropy
entro = -sum( probs.*log2(probs+eps) );

title([ 'Entropy = ' num2str(entro) ])

%% done.
